lib/querytables.js

```diff
@@ -1,44 +1,280 @@
+            SELECT
+                quote_ident(n.nspname::text) schema_name,
+                quote_ident(c.relname::text) table_name,
+                c.relkind,
+                cdb_table_oids.*,
+                (SELECT md.updated_at FROM cartodb.CDB_TableMetadata md WHERE md.tabname = reloid) AS updated_at,
+                (CASE   WHEN relkind != 'f' THEN current_database()
+                        ELSE (
+                            SELECT option_value AS dbname FROM cdb_table_oids, pg_options_to_table((
+                                SELECT fs.srvoptions
+                                FROM pg_foreign_table ft
+                                LEFT JOIN pg_foreign_server fs ON ft.ftserver = fs.oid
+                                WHERE ft.ftrelid = cdb_table_oids.reloid
+                            )) WHERE option_name='dbname')
+                         END) AS dbname
+            FROM cdb_table_oids, pg_catalog.pg_class c
+            LEFT JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid
+            WHERE c.oid = cdb_table_oids.reloid
+        )
+        SELECT * FROM cdb_table_metadata ORDER BY reloid;`;
 
-    pg.query(query, (err, result) => {
+    const readOnly = true;
+    pg.query(metadataQuery, {}, (err, result) => {
         if (err) {
             const msg = err.message ? err.message : err;
-            return callback(new Error('could not fetch affected tables or last updated time: ' + msg));
+            return callback(new Error('Could not fetch metadata about the affected tables: ' + msg));
         }
-        result = result || {};
         const rows = result.rows || [];
+        const getForeignTableUpdatedAtPromise = promisify(getForeignTableUpdatedAt);
+
+        /* For foreign tables we need to do extra queries to extract the updated_at properly */
+        const promiseCollection = rows
+                .filter(r => r.relkind === 'f')
+                .map(row => getForeignTableUpdatedAtPromise(pg, row)
+                    .then(updated_at => Object.assign(row, {updated_at})));
+        Promise.all(promiseCollection)
+            .then(() => callback(null, rows))
+            .catch(err => callback(err));
+    }, readOnly);
+}
+
+/**
+ * Returns a DatabaseTables Object that includes the information about the tables
+ * affected by a query (as seen by the planner)
+ * @param {Object} pg         - Database connection (PSQL)
+ * @param {String} sql        - Database query
+ * @param {Function} callback - Cb function ({Error}, {Object::DatabaseTablesEntry})
+ */
+module.exports.getQueryMetadataModel = function (pg, sql, callback) {
+    const iSQL = replaceTile0(sql);   
+    
+    const statements = getQueryStatements(iSQL);
+    const getAffectedTablesPromise = promisify(getAffectedTables);
+    const promiseCollection = statements.map(query => getAffectedTablesPromise(pg, query));
+
+    Promise.all(promiseCollection)
+        .then((values) => {
+            const merged = [].concat.apply([], values);
```

Comment:
> Or even better, return the actual value in getAffectedTablesPromise instead of the array which is just wrapping the value that we want to use

Each call to getAffectedTables returns an array because it might return multiple strings, not just one, so unless you are suggesting returning one element when possible and an array when there are multiple of them (and I hope you aren't), that's not possible.

Reply:
Ok, I thought it returns one `string`.